Api/get end point to fetch all historical data

BonfaceKilz opened the pull request:
# Description
This PR adds an extra end-point to fetch all the RIF metadata.  In addition, this PR also fixes the GET wiki endpoint to use the correct header to fetch json-ld data.

# Testing

```
╭─munyoki@saitama ~  
╰─➤  curl localhost:8080/api/metadata/wiki/1/history
[
  {
    "comment": "hippocampal CA1-3 expression signature", 
    "created": "2005-12-17T00:23:23", 
    "email": "rwilliam@nb.utmem.edu", 
    "pubmed_id": "15084669", 
    "species": "mouse", 
    "symbol": "Lpl", 
    "type": "gnc:GNWikiEntry", 
    "version": "0", 
    "wiki_identifier": "gn:wiki-1-0"
  }
]

curl localhost:8080/api/metadata/wiki/1/history --header "Accept: application/ld+json"

{
  "@context": {
    "categories": "gnt:belongsToCategory", 
    "comment": "rdfs:comment", 
    "created": "dct:created", 
    "data": "@graph", 
    "dct": "http://purl.org/dc/terms/", 
    "email": "foaf:mbox", 
    "foaf": "http://xmlns.com/foaf/0.1/", 
    "gn": "http://genenetwork.org/id/", 
    "gnc": "http://genenetwork.org/category/", 
    "gnt": "http://genenetwork.org/term/", 
    "id": "dct:identifier", 
    "initial": "gnt:initial", 
    "pubmed_id": "dct:references", 
    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#>", 
    "rdfs": "http://www.w3.org/2000/01/rdf-schema#", 
    "reason": "gnt:reason", 
    "species": "gnt:species", 
    "symbol": "rdfs:label", 
    "type": "@type", 
    "version": "gnt:hasVersion", 
    "web_url": "foaf:homepage", 
    "wiki_identifier": "@id"
  }, 
  "data": [
    {
      "comment": "hippocampal CA1-3 expression signature", 
      "created": "2005-12-17T00:23:23", 
      "email": "rwilliam@nb.utmem.edu", 
      "pubmed_id": "15084669", 
      "species": "mouse", 
      "symbol": "Lpl", 
      "type": "gnc:GNWikiEntry", 
      "version": "0", 
      "wiki_identifier": "gn:wiki-1-0"
    }
  ]
}

```

BonfaceKilz committed:
Restructure RDF module.

* gn3/api/metadata.py: Import constants from gn3.db.rdf
* gn3/api/metadata_api/wiki.py: Ditto.  Import
"get_wiki_entries_by_symbol" from gn3.db.rdf.wiki.
* gn3/db/constants.py: Delete file and move all constants ...
* gn3/db/rdf.py: ... and functions ...
* gn3/db/rdf/__init__.py: ... here.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Add missing doc-strings to make pylint pass.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Ignore type error and add a FIXME stub.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Create global wiki context usable by all functions in gn3.rdf.wiki

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Use correct prefix when querying for email in triple.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Always output the top-level graph when compacting.

This makes sure that the API is always consistent.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Add comment history.

* gn3/api/metadata_api/wiki.py: Import get_comment_history.
(get_history): New end-point.
* gn3/db/rdf/wiki.py: (get_comment_history): New function.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Use the correct header to request json-ld data.

In HTTP, we use content negotiation to server different
representations of a resource to the same URI to help the user agent
specify which representation is best suited for the user:
    <https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation>
The "Accept" requet HTTP indicates which content types the user is
supposed to understand:
    <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept>

* gn3/api/metadata_api/wiki.py (get_wiki_entries): Use the "Accept"
header.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Fix type error.

* gn3/api/metadata_api/wiki.py: Remove FIXME comment.
* gn3/db/wiki.py: Fix type error.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

jnduli commented:
Here's what I expected the data section to contain:
```
{
  "categories": [
    "Behavior", 
    "Biochemistry"
  ], 
  "comment": "XXXX", 
  "email": "XXXX", 
  "initial": "XXXX", 
  "pubmed_ids": [
    "XXXX", 
    "XXXX"
  ], 
  "reason": "XXXX", 
  "species": "mouse", 
  "symbol": "XXXX", 
  "version": 14, 
  "weburl": "XXXX"
}
```
The diffs I see are:

- we don't have multiple pubmed_ids.
- we don't have categories, initial, weburl and reason
- version we return in a string instead of int
- we have an extra wiki_identifier(which I understand), but what's the purpose for type?

BonfaceKilz committed:
Remove type from wiki history.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Rename "pubmed_id" -> "pubmed_ids" to be explicit it's an array.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Set OPTIONAL values to default to an empty string.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Enforce array type to "categories" and "pmids" fields.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Sort the RIF history data in reverse order.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz committed:
Correctly get categories in response.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

BonfaceKilz commented:
@jnduli addressed above comments:

```
╭─munyoki@saitama ~/projects/genenetwork3  ‹api/get-end-point-to-fetch-all-historical-data*› 
╰─➤  curl localhost:8080/api/metadata/wiki/587/history 
[
  {
    "categories": [
      "Physiology and function", 
      "Epigenetic modulation", 
      "Expression patterns: temporal variation"
    ], 
    "comment": "Imprinted in mouse and human. Peg3 is a paternally expressed, zinc finger protein. Peg3 is expressed in early somites, branchial arches and other mesodermal tissues, as well as in the hypothalamus. Loss of Peg3 expression results in reduced maternal care, maternal food intake, milk let-down, placental size, impaired fetal growth, suckling and postnatal growth, and delayed puberty. from Morison", 
    "created": "2010-06-24T04:32:53", 
    "email": "rwilliam@nb.uthsc.edu", 
    "initial": "RWW", 
    "pubmed_ids": [
      7550314, 
      15306355
    ], 
    "reason": "more", 
    "species": "mouse", 
    "symbol": "Peg3", 
    "version": 0, 
    "web_url": "http://igc.otago.ac.nz/Search.html", 
    "wiki_identifier": "gn:wiki-587-0"
  }, 
  {
    "categories": [
      "Physiology and function", 
      "Epigenetic modulation", 
      "Expression patterns: temporal variation"
    ], 
    "comment": "Peg3 is a paternally expressed, zinc finger protein. Peg3 is expressed in early somites, branchial arches and other mesodermal tissues, as well as in the hypothalamus. Loss of Peg3 expression results in reduced maternal care, maternal food intake, milk let-down, placental size, impaired fetal growth, suckling and postnatal growth, and delayed puberty. \nGlasgow et al, 2005 have described a Peg3 antisense transcript in rat APeg3. Its imprinting status is not yet known.", 
    "created": "2006-10-31T20:42:43", 
    "email": "rwilliam@nb.utmem.edu", 
    "initial": "RWW", 
    "pubmed_ids": [
      15306355, 
      7550314
    ], 
    "reason": "", 
    "species": "mouse", 
    "symbol": "Peg3", 
    "version": 1, 
    "web_url": "", 
    "wiki_identifier": "gn:wiki-587-1"
  }
]
```

jnduli commented on gn3/db/rdf/wiki.py:
```diff
@@ -0,0 +1,164 @@
+        else:
+            result["pubmed_ids"] = []
+        result["version"] = int(result["version"])
+    # We manually sort the array, since somehow "ORDER BY" in the
```
should we also mention this in the docs for the function in question?

jnduli commented on gn3/db/rdf/wiki.py:
```diff
@@ -0,0 +1,164 @@
+    # sparql query is jumbled up during framing and compacting.  This
+    # operation isn't heavy since we don't get many versions per wiki
+    # entry.
+    results["data"] = sorted(data, key=lambda d: d["created"], reverse=True)
```
The `createtime` column in our db defaults to `0000-00-000`. I suggest we change the key to `(d['version'], d['created'])` instead

BonfaceKilz replied:
I deleted this commit and did a force push.

BonfaceKilz committed:
Update comment explaining why CONSTRUCT doesn't respect ORDER BY.

Signed-off-by: Munyoki Kilyungi <me@bonfacemunyoki.com>

jnduli commented on gn3/db/rdf/wiki.py:
```diff
@@ -0,0 +1,168 @@
+    # CONSTRUCT. Using ORDER BY on a solution sequence for a CONSTRUCT
+    # or DESCRIBE query has no direct effect because only SELECT
+    # returns a sequence of results.  See:
+    # <https://stackoverflow.com/questions/78186393/how-apply-an-order-to-construct-in-sparql#:~:text=SPARQL%20engine%20does%20respect%20such,is%20a%20set%20of%20triples.>
```
can we replace this with: https://stackoverflow.com/q/78186393

BonfaceKilz replied:
Done.
